client/src/main/java/io/pravega/client/netty/impl/ClientConnectionInboundHandler.java

```diff
@@ -167,28 +169,42 @@ public void close() {
         }
     }
 
+    private Channel getChannel() throws ConnectionFailedException {
+        Channel ch = channel.get();
+        if (ch == null) {
+            throw new ConnectionFailedException("Connection to " + connectionName + " is not established.");
+        }
+        return ch;
+    }
+
     /**
-     * This method ensures io.pravega.client.netty.impl.ClientConnectionInboundHandler#channelActive(io.netty.channel.ChannelHandlerContext)
-     * is invoked before attempting to return the channel.
+     * This function completes the input future when the change is registered.
      *
-     * @return io.netty.channel.Channel
-     * @throws ConnectionFailedException incase channelActive() is not invoked or channel is null due to
-     * channelInactive() invocation.
+     * @param future CompletableFuture which will be completed once the channel is registered.
      */
-    private Channel getChannel() throws ConnectionFailedException {
-        try {
-            Exceptions.handleInterrupted(() -> channelRegisteredLatch.await(CHANNEL_ACTIVE_TIMEOUT));
-            Channel ch = channel.get();
-            if (ch == null) {
-                throw new ConnectionFailedException("Connection to " + connectionName + " is not established.");
-            }
-            return ch;
-        } catch (TimeoutException e) {
-            throw new ConnectionFailedException("Channel is not active. Connection to " + connectionName + "is not " +
-                    "established.");
+    void completeWhenRegistered(final CompletableFuture<Void> future) {
+        if (isRegistered.get()) { // channelRegistered is already invoked, complete the future.
+            future.complete(null);
+        } else {
+            BinaryOperator<CompletableFuture<Void>> accumulate = (prev, current) -> {
```

Comment:
Should we do a private method rather than declaring it like this for clarity?

Reply:
done, refactored the code.